src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java

```diff
@@ -75,45 +88,127 @@ FakeWither getWither(Player player, String message) {
+        }
+        else {
+            return player.getLocation().add(player.getEyeLocation().getDirection().multiply(110));
+        }
+    }
+
+    private void destroy(Player player) {
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            FakeWither wither = players.get(player);
+            if (wither == null)
+                return;
+            ((CraftPlayer) player).getHandle().playerConnection.sendPacket(wither.getDestroyPacket());
+            players.remove(player);
+        }
+    }
+
+    @EventHandler(priority = EventPriority.MONITOR)
+    public void onPlayerJoin(PlayerJoinEvent event) {
+        addWither(event.getPlayer(), "", false);
+        if (TimeLimit.getMatchTimeLimit() == 0) {
+            BossBar.delete();
+        }
+    }
+
+    @EventHandler(priority = EventPriority.MONITOR)
+    public void onCycleComplete(CycleCompleteEvent event) {
+        for (Player player : Bukkit.getOnlinePlayers()) {
+            addWither(player, "", false);
+        }
+        if (TimeLimit.getMatchTimeLimit() == 0) {
+            BossBar.delete();
+        }
     }
 
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
-    public void PlayerLogout(PlayerQuitEvent event) {
+    public void onPlayerLogout(PlayerQuitEvent event) {
         players.remove(event.getPlayer());
     }
 
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onPlayerMove(final PlayerMoveEvent event) {
-        handleTeleport(event.getPlayer(), event.getTo());
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            if (players.get(event.getPlayer()) != null) {
+                handleTeleport(event.getPlayer(), event.getTo(), true);
+            }
+        }
+        if (TimeLimit.getMatchTimeLimit() == 0) {
+            BossBar.delete();
+        }
     }
 
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onPlayerTeleport(final PlayerTeleportEvent event) {
-        handleTeleport(event.getPlayer(), event.getTo());
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            if (players.get(event.getPlayer()) != null) {
+                handleTeleport(event.getPlayer(), event.getTo(), true);
+            }
+        }
+        if (TimeLimit.getMatchTimeLimit() == 0) {
+            BossBar.delete();
+        }
     }
 
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onPlayerRespawn(final PlayerRespawnEvent event) {
-        handleTeleport(event.getPlayer(), event.getRespawnLocation().clone());
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            if (players.get(event.getPlayer()) != null) {
+                handleTeleport(event.getPlayer(), event.getRespawnLocation().clone(), true);
+            }
+
+        }
+        if (TimeLimit.getMatchTimeLimit() == 0) {
```

Comment:
@Alfalik Your if statement checking if the match time limit is zero and then deleting the boss bar is a bit repetitious, you should consolidate it down into one statement in the handleTeleport method.